# Homework 5

## Q1
1. Compute the gradient $\nabla f(x)$ and Hessian $\nabla^{2} f(x)$ of the Rosenbrock
function

\begin{aligned}
f\left(x_{1}, x_{2}\right)=100\left(x_{2}-x_{1}^{2}\right)^{2}+\left(1-x_{1}\right)^{2}
\end{aligned}

## Q1 Solution

The gradient has the following form:

\begin{equation}
\nabla f=\left[\begin{array}{c}
\frac{\partial f}{\partial x_{1}} \\
\frac{\partial f}{\partial x_{2}}
\end{array}\right]
\end{equation}

Let's calculate the derivative with respect to $x_1$ first:

\begin{aligned}
&\frac{\partial f}{\partial x_{1}}=200\left(x_{2}-x_{1}^{2}\right)\left(-2 x_{1}\right)-2\left(1-x_{1}\right) \\
&\frac{\partial f}{\partial x_{1}}=-2\left[\left(200 x_{2}-200 x_{1}^{2}\right)\left(x_{1}\right)+1-x_{1}\right] \\
&\frac{\partial f}{\partial x_{1}}=-2\left(200 x_{1} x_{2}-200 x_{1}^{3}+1-x_{1}\right) \\
&\frac{\partial f}{\partial x_{1}}=2\left(200 x_{1}^{3}-200 x_{1} x_{2}+x_{1}-1\right)
\end{aligned}

Then, we calculate the derivative with respect to $x_2$:

\begin{equation}
\frac{\partial f}{\partial x_{2}}=200\left(x_{2}-x_{1}^{2}\right)
\end{equation}

Thus, the gradient $\nabla f(x)$ is:

\begin{equation}
\nabla f=\left[\begin{array}{c}
2\left(200 x_{1}^{3}-200 x_1 x_2+x_{1}-1\right) \\
200\left(x_2-x_{1} 2\right)
\end{array}\right]
\end{equation}

We will use the result above to derive the Hessian matrix which has the following form:

\begin{equation}
\textbf H f=\left[\begin{array}{ll}
\frac{\partial^{2} f}{\partial x_{1}^{2}} & \frac{\partial^{2} f}{\partial x_{1} \partial x_{2}} \\
\frac{\partial^{2} f}{\partial x_2 \partial x_{1}} & \frac{\partial^{2} f}{\partial x_{2}^{2}}
\end{array}\right]
\end{equation}

Using the previous results we have:

\begin{equation}
\begin{aligned}
&\frac{\partial^{2} f}{\partial x_{1}^{2}}=1200 x_{1}^{2}-400 x_{2}+2 \\
&\frac{\partial^{2} f}{\partial x_{2}^{2}}=200 \\
&\frac{\partial^{2} f}{\partial x_{1} \partial x_{2}}=-400 x_{1} \\
&\frac{\partial^{2} f}{\partial x_{2} \partial x_{1}}=-400 x_{1}
\end{aligned}
\end{equation}

And thus the Hessian matrix is:

\begin{equation}
\textbf H f=\left[\begin{array}{ll}
1200x_1^2-400x_2 +2 & -400 x_1 \\
-400 x_1 & 200
\end{array}\right]
\end{equation}

## Q2
Implement the Newton’s Method with line search given in Algorithm 1. Use the Newton’s
method to minimize the Rosenbrock function in Problem 1. Set the initial stepsize
$\bar{\alpha}=1$ Select your own choice of $\rho \in(0,1), c \in(0,1)$. First run the
algorithm from the initial point $x^{0}=(1.2,1.2)^{\top}$, and then try the more
difficult starting point $x^{0}=(-1.2,1)^{\top}$. For each starting point, print out
the step length $\alpha^{k}$ used by the algorithm as well as the point $x^{k}$ for
every step $k$. You should observe that Newton's Method converges very fast.

![Algorithm 1](algo_1.jpg)

## Q2 Solution
First, let's implement the algorithm using python and the numpy package. We will use
the results from Q1 to define the gradient and hessian of the Rosenbrock function.

In [45]:
import numpy as np

def f(x):
    return 100 * (x[1] - x[0] ** 2) ** 2 + (1 - x[0]) ** 2

def df(x):
    nabla = np.zeros(2)
    nabla[0] = 2 * (200 * x[0] ** 3 - 200 * x[0] * x[1] + x[0] - 1)
    nabla[1] = 200 * (x[1] - x[0] ** 2)

    return nabla

def d2f(x):
    hess = np.zeros((2, 2))
    hess[0, 0] = 1200 * x[0] ** 2 - 400 * x[1] + 2
    hess[1, 1] = 200
    hess[0, 1] = hess[1, 0] = -400 * x[0]

    return hess

def min_rosenbrock(x_0, alpha_bar, eps):
    k = 0
    d_0 = -np.linalg.inv(d2f(x_0)) @ df(x_0)
    x = x_0

    d = d_0
    while np.linalg.norm(df(x)) > eps:
        alpha = alpha_bar
        rho = np.random.rand()
        c = np.random.rand()
        while f(x + alpha * d) > f(x) + c * alpha * df(x).T @ d:
            alpha = rho * alpha
        print(f"----Iteration {k}----")
        print(f"Step Length: {alpha}")
        print(f"x^{k}: {x}")
        x = x + alpha * d
        d = -np.linalg.inv(d2f(x)) @ df(x)
        k += 1

In [46]:
x_0 = np.array([1.2, 1.2])
alpha_bar = 1
eps = 10 ** -4

min_rosenbrock(x_0, alpha_bar, eps)

----Iteration 0----
Step Length: 1
x^0: [1.2 1.2]
----Iteration 1----
Step Length: 0.5346613079766228
x^1: [1.19591837 1.43020408]
----Iteration 2----
Step Length: 0.4385207573079363
x^2: [1.09151626 1.18050019]
----Iteration 3----
Step Length: 1
x^3: [1.07890219 1.15774647]
----Iteration 4----
Step Length: 0.0020247873618312916
x^4: [1.0439386  1.08858535]
----Iteration 5----
Step Length: 1
x^5: [1.04386711 1.08843857]
----Iteration 6----
Step Length: 0.07266585425150308
x^6: [1.00860406 1.01603867]
----Iteration 7----
Step Length: 0.6826994266001746
x^7: [1.00810336 1.01511902]
----Iteration 8----
Step Length: 1
x^8: [1.00360814 1.00684312]
----Iteration 9----
Step Length: 0.1577285736172359
x^9: [1.00025869 1.00050623]
----Iteration 10----
Step Length: 1
x^10: [1.00021798 1.00042656]


From the above result we can see that it took 7 iterations to reach the optimum solution
of $x = [1, 1]^T$ with the minimum value being zero, starting from the point
$x^0 = [1.2 , 1.2]^T$.

Next, we will start the optimization from the point $x^0 = [-1.2, 1]^T$

In [47]:
x_0 = np.array([-1.2, 1])
min_rosenbrock(x_0, alpha_bar, eps)

----Iteration 0----
Step Length: 0.14148148058839893
x^0: [-1.2  1. ]
----Iteration 1----
Step Length: 0.47124618019569325
x^1: [-1.1965027   1.05385834]
----Iteration 2----
Step Length: 1
x^2: [-1.18298128  1.19951963]
----Iteration 3----
Step Length: 0.13335976773408106
x^3: [-1.12971836  1.27342664]
----Iteration 4----
Step Length: 0.011396815034403556
x^4: [-0.94851318  0.86438334]
----Iteration 5----
Step Length: 0.1407139139432948
x^5: [-0.94575758  0.85955812]
----Iteration 6----
Step Length: 0.19900112491131183
x^6: [-0.91144678  0.79956955]
----Iteration 7----
Step Length: 0.2991597819906734
x^7: [-0.85885826  0.70990828]
----Iteration 8----
Step Length: 1
x^8: [-0.77390431  0.57227694]
----Iteration 9----
Step Length: 1
x^9: [-0.49367498  0.1651865 ]
----Iteration 10----
Step Length: 0.35898899711202803
x^10: [-0.40426386  0.15543492]
----Iteration 11----
Step Length: 0.27096669509823623
x^11: [-0.21028904  0.00147079]
----Iteration 12----
Step Length: 1
x^12: [-0.17594943 -0

In the last case, it took 27 iterations to reach the optimum solution.

# Q3

Figure 1 below illustrates the water network of Newvillage. The lines are water
piplelines numbered from 1 through 13 . The arrows on the lines are possible
direction(s) of flow of water in these pipelines. The circles are water sources
numbered $A, B, C$. The rectangles are houses $\mathrm{D}, \mathrm{E}, \mathrm{F},
\mathrm{G}, \mathrm{H}$. The maximum possible capacity of the water sources are (the
sources can operate at less than the maximum capacity): A: 100 Units, B: 100 Units, C:
80 Units Demands of water in the houses are: D: 50 Units, E: 60 Units, F: 40 Units, G:
30 Units, H: 70 Units Since the houses are at different elevation and the pipes are of
different diameter, the cost of transporting water is different in the different pipes.
These costs per unit of water are: Pipe 1: $\$ 2$, Pipe 2: $\$ 3$, Pipe $3: \$ 4$, Pipe
4: $\$ 2$, Pipe 5: $\$ 3$, Pipe 6: $\$ 2$, Pipe 7: $\$ 4$, Pipe 8: $\$ 1$, Pipe 9:
$\$ 2$, Pipe 10: $\$ 4$, Pipe 11: $\$ 5$, Pipe 12: $\$ 1$, Pipe 13: $\$ 2$. Formulate
an LP to minimize the total cost of transporting water so as to meet the water demands
of each house.

![Water Network](water_net_q2.png)

## Q3 Solution
We will focus this formulation using the arcs between the sources and the houses to
define the objective function and constraints, mostly using the flow conservation
principle between the nodes (sources and houses).

The set of Arcs $A$ is defined by the letters of the sources and houses, i.e: the arc
between the source A and the house D is **AD**

The **objective function** to minimize is the following:

\begin{aligned}
&\min 2 x_{A D}+4 x_{A E}+2 x_{D F}+4 x_{E F}+2 x_{EH} + 4 x_{HF}+3 x_{B D}+3 x_{B F}\\
&+2 x_{F D}+2 x_{BG}+5 x_{C F}+2 x_{C H}+x_{C G}+x_{F G}+2x_{H E}
\end{aligned}

> Notice that we are including the quantities of water that can go in another direction
as additional variables since ($x_{DF}, x_{EH}$) are equivalent in cost to ($x_{FD},
x_{HE}$). We need to define them as separate variables as they will have a different
sign in the constraints related to the flow conservation principle as formulated below.
> However, it is expected that the solver would choose only one of the equivalent
> variables with a value higher than zero.

**Program Constraints**:

\begin{aligned}
&x_{A D}+x_{A E} \leq 100 \\
&x_{B D}+x_{B F}+x_{B G} \leq 100 \\
&x_{CH}+x_{C F}+x_{C G} \leq 80 \\
\\
&x_{A D}+x_{B D}+x_{F D}-x_{D F} \geq 50\\
&x_{A E}-x_{E F}+x_{H E}-x_{E H} \geq 60\\
&x_{D F}-x_{F D}+x_{BG}-x_{FG}+x_{C F}+x_{H F}+x_{E F} \geq 40\\
&x_{B G}+x_{F G}+x_{C G} \geq 30\\
&x_{E H}-x_{H E}-x_{H F}+x_{CH} \geq 70 \\
\\
&x_{i j} \geq 0 \quad \forall(i, j) \in A
\end{aligned}

## Q4

Consider the following electric power network shown in Figure 2. This network is taken
from a real-world electric power system. Electricity generators are located at
nodes 1, 3, and 5 and producing $p_{1}, p_{2}, p_{3}$ amounts of electricity,
respectively. Electricity loads are located at nodes 2, 4 , and 6 and are consuming
$d_{1}, d_{2}, d_{3}$ amounts of electricity, respectively. The demand is fixed and
given as $d_{1}=110, d_{2}=65, d_{3}=95$.
Each generator $i$ 's production must be within an upper and a lower bound as
$p_{i}^{\min } \leq p_{i} \leq p_{i}^{\max }$. The bounds are given as
$p_{1}^{\min }=20, p_{1}^{\max }=200, p_{2}^{\min }=20, p_{2}^{\max }=150,
p_{3}^{\min }=10, p_{3}^{\max }=$ $150 .$
The flow limits over lines are given as $f_{12}^{\max }=100, f_{23}^{\max }=110,
f_{34}^{\max }=50, f_{45}^{\max }=80, f_{56}^{\max }=$ $60, f_{61}^{\max }=40$
The line parameters are given as $B_{12}=11.6, B_{23}=5.9, B_{34}=13.7, B_{45}=9.8,
B_{56}=$ $5.6, B_{61}=10.5 .$ The unit generation costs are given as $c_{1}=16,
c_{2}=20, c_{3}=8$.

(a) Formulate the power system scheduling problem using the model discussed in
Lecture $2 .$

(b) Implement and solve the model using CVXPY. Write down the optimal solution.

(c) Find the electricity prices for demand at nodes $2,4,6$. To do this, use the
command constraints[0].dual_value to find the dual variable of constraints[0].
Hint: Recall the electricity price at node $i$ is the dual variable for the flow
conservation constraint at node $i .$

![Electric Network](electric_network.png)

## Q4 Solution
---
### Q4.a Solution
First, we define the cost function:

$\min 10 p_1 + 20 p_2 + 8 p_3$

Then, we have the following flow definitions to define the problem constraints:

![User defined flows](electric_network_flows.png)

**Constraints:**
- Flow Constraints

\begin{array}{ll}
f_{12}+f_{16}=p_{1} & f_{12}+f_{32}=110 \\
f_{32}+f_{34}=p_{2} & f_{34}+f_{5 4}=65 \\
f_{5 6}+f_{54}=p_{3} & f_{16}+f_{56}=95
\end{array}

- Branch flow and nodal potential

\begin{aligned}
&f_{12}=11.6\left(\theta_{1}-\theta_{2}\right) \\
&f_{16}=10.5\left(\theta_{1}-\theta_{6}\right) \\
&f_{32}=5.9\left(\theta_{3}-\theta_{2}\right) \\
&f_{34}=13.7\left(\theta_{3}-\theta_{4}\right) \\
&f_{5 6}=5.6\left(\theta_{5}-\theta_{6}\right) \\
&f_{54}=9.8\left(\theta_{5}-\theta_{4}\right)
\end{aligned}

- Flow limit constraints

\begin{aligned}
-100 \leq &f_{12} \leq 100 \\
-40 \leq &f_{16} \leq 40 \\
-110 \leq &f_{32}  \leq 110 \\
-50 \leq &f_{34}  \leq 50 \\
-60 \leq &f_{56}  \leq 60 \\
-80 \leq &f_{54} \leq 80
\end{aligned}

- Generator physical limit constraints

\begin{aligned}
&20 \leq p_{1} \leq 200 \\
&20 \leq p_{2} \leq 150 \\
&10 \leq P_{3} \leq 150
\end{aligned}

### Q4.b Solution

In [48]:
import cvxpy as cp

# Define variables
p1, p2, p3 = cp.Variable(), cp.Variable(), cp.Variable()

f12, f16, f32, f34, f56, f54 = cp.Variable(), cp.Variable(), cp.Variable(), \
                               cp.Variable(), cp.Variable(), cp.Variable()

theta1, theta2, theta3, theta4, theta5, theta6 = cp.Variable(), cp.Variable(), \
                                                 cp.Variable(), cp.Variable(), \
                                                 cp.Variable(), cp.Variable()

# Objective function
obj = cp.Minimize(10 * p1 + 20 * p2 + 8 * p3)

constraints = [
    # Flow Constraints (Generators)
    f12 + f16 == p1,
    f32 + f34 == p2,
    f56 + f54 == p3,
    # Flow Constraints (Demand)
    f12 + f32 == 110,
    f34 + f54 == 65,
    f16 + f56 == 95,
    # Branch flow and nodal potential constraints
    f12 == 11.6 * (theta1 - theta2),
    f16 == 10.5 * (theta1 - theta6),
    f32 == 5.9 * (theta3 - theta2),
    f34 == 13.7 * (theta3 - theta4),
    f56 == 5.6 * (theta5 - theta6),
    f54 == 9.8 * (theta5 - theta4),
    # Flow limit constraints
    -100 <= f12, f12 <= 100,
    -40 <= f16, f16 <= 40,
    -110 <= f32, f32 <= 110,
    -50 <= f34, f34 <= 50,
    -60 <= f56, f56 <= 60,
    -80 <= f54, f54 <= 80,
    # Generator physical limita constraints
    20 <= p1, p1 <= 200,
    20 <= p2, p2 <= 150,
    10 <= p3, p3 <= 150,
]

# Problem definition
prob = cp.Problem(obj, constraints)

# Little helper method to get the variable values rounded to two decimals
def r(x):
    return round(float(x.value), 2)

In [49]:
print("Optimal value: ", round(prob.solve()))
print("----------------------------------\n")
print(f"Variable results:\n")
print(f"p1: {r(p1)}, p2: {r(p2)}, p3: {r(p3)}")
print(f"f12: {r(f12)}, f16: {r(f16)}, f32: {r(f32)}, "
      f"f34: {r(f34)}, f54: {r(f54)}, f56: {r(f56)}")
print(f"theta1: {r(theta1)}, theta2: {r(theta2)}, theta3: {r(theta3)}, "
      f"theta4: {r(theta4)}, theta5: {r(theta5)}, theta6: {r(theta6)}")

Optimal value:  2626
----------------------------------

Variable results:

p1: 113.12, p2: 20.0, p3: 136.88
f12: 78.12, f16: 35.0, f32: 31.88, f34: -11.88, f54: 76.88, f56: 60.0
theta1: 0.92, theta2: -5.81, theta3: -0.41, theta4: 0.46, theta5: 8.3, theta6: -2.41


From the output above we can that the **Minimum price value** is **2626**

### Q4.c Solution
Here we get the duals from the flow constraints in the sinks (consumers). This also
represents the Locational Marginal Price. Recall that these come from the fourth to
the sixth constraints.

In [50]:
print("Marginal Price Location 2: ", -prob.constraints[3].dual_value)
print("Marginal Price Location 4: ", -prob.constraints[4].dual_value)
print("Marginal Price Location 6: ", -prob.constraints[5].dual_value)

Marginal Price Location 2:  9.599719042811598
Marginal Price Location 4:  8.473801950978045
Marginal Price Location 6:  10.442215155064769
